model: Add SauerkrautLM-ColPali visual document retrieval models

dgolchin opened the pull request:
Add inference code and requirements for SauerkrautLM-ColPali visual document retrieval models.

These are multi-vector embedding models based on the ColPali architecture:
- ColQwen3 (Qwen3-VL backbone): 1.7B Turbo, 2B, 4B, 8B variants
- ColLFM2 (LFM2-VL backbone): 450M variant
- ColMinistral3 (Ministral3 backbone): 3B variant

All models produce 128-dimensional embeddings per text/image token and use MaxSim (late interaction) for retrieval scoring.

Model checkpoints:
- https://huggingface.co/VAGOsolutions/SauerkrautLM-ColQwen3-1.7b-Turbo-v0.1
- https://huggingface.co/VAGOsolutions/SauerkrautLM-ColQwen3-2b-v0.1
- https://huggingface.co/VAGOsolutions/SauerkrautLM-ColQwen3-4b-v0.1
- https://huggingface.co/VAGOsolutions/SauerkrautLM-ColQwen3-8b-v0.1
- https://huggingface.co/VAGOsolutions/SauerkrautLM-ColLFM2-450M-v0.1
- https://huggingface.co/VAGOsolutions/SauerkrautLM-ColMinistral3-3b-v0.1

If you add a model or a dataset, please add the corresponding checklist:

* [dataset checklist](https://embeddings-benchmark.github.io/mteb/contributing/adding_a_dataset/#submit-a-pr)
* [model checklist](https://embeddings-benchmark.github.io/mteb/contributing/adding_a_model/#submitting-your-model-as-a-pr)

Samoed commented on mteb/models/model_implementations/slm_models.py:
```diff
@@ -0,0 +1,484 @@
+    return SLMColLFM2Wrapper(model_name=model_name, revision=revision, device=device, **kwargs)
+
+def slm_colministral3_loader(model_name: str, revision: str | None = None, device: str | None = None, **kwargs) -> SLMColMinistral3Wrapper:
+    return SLMColMinistral3Wrapper(model_name=model_name, revision=revision, device=device, **kwargs)
```
You can you class directly
```suggestion
```

Samoed commented on mteb/models/model_implementations/slm_models.py:
```diff
@@ -0,0 +1,484 @@
+        )
+        
+        self.device = device or ("cuda" if torch.cuda.is_available() else "cpu")
+        self._load_model_and_processor(model_name, revision, use_flash_attn, **kwargs)
```
Can you change to something like? This would simplify code a bit
```suggestion
        self._load_model_and_processor(model_name, revision, use_flash_attn, **kwargs)
        self.mdl = self.mdl.to(self.device)
        self.mdl.eval()
```

Samoed commented on mteb/models/model_implementations/slm_models.py:
```diff
@@ -0,0 +1,449 @@
+
+# ColLFM2-450M: ~450M params → 900 MB VRAM in bfloat16
+slm_collfm2_450m = ModelMeta(
+    loader=partial(SLMColLFM2Wrapper),
```
```suggestion
    loader=SLMColLFM2Wrapper,
```

Samoed commented on mteb/models/model_implementations/slm_models.py:
```diff
@@ -0,0 +1,449 @@
+
+# ColQwen3-8B: ~8B params → 16 GB VRAM in bfloat16
+slm_colqwen3_8b = ModelMeta(
+    loader=partial(SLMColQwen3Wrapper),
```
```suggestion
    loader=SLMColQwen3Wrapper,
```

Samoed commented on mteb/models/model_implementations/slm_models.py:
```diff
@@ -0,0 +1,449 @@
+
+# ColQwen3-2B: ~2.2B params → 4.4 GB VRAM in bfloat16
+slm_colqwen3_2b = ModelMeta(
+    loader=partial(SLMColQwen3Wrapper),
```
```suggestion
    loader=SLMColQwen3Wrapper,
```

Samoed commented on mteb/models/model_implementations/slm_models.py:
```diff
@@ -0,0 +1,449 @@
+
+# ColQwen3-1.7B Turbo: ~1.7B params → 3.4 GB VRAM in bfloat16
+slm_colqwen3_1_7b_turbo = ModelMeta(
+    loader=partial(SLMColQwen3Wrapper),
```
```suggestion
    loader=SLMColQwen3Wrapper,
```

Samoed commented on mteb/models/model_implementations/slm_models.py:
```diff
@@ -0,0 +1,449 @@
+    loader=partial(SLMColQwen3Wrapper),
+    name="VAGOsolutions/SauerkrautLM-ColQwen3-1.7b-Turbo-v0.1",
+    languages=SUPPORTED_LANGUAGES,
+    revision="main",
```
Can you specify exact revisions of your models?

Samoed commented on mteb/models/model_implementations/slm_models.py:
```diff
@@ -0,0 +1,449 @@
+    similarity_fn_name=ScoringFunction.MAX_SIM,
+    use_instructions=True,
+    training_datasets=None,
+    citation=SAUERKRAUTLM_CITATION + COLPALI_CITATION,
```
Can you add base models for all your models?
```suggestion
    citation=SAUERKRAUTLM_CITATION + COLPALI_CITATION,
    adapted_from="Qwen/Qwen3-VL-2B",
```

Samoed commented on mteb/models/model_implementations/slm_models.py:
```diff
@@ -0,0 +1,449 @@
+    reference="https://huggingface.co/VAGOsolutions/SauerkrautLM-ColQwen3-1.7b-Turbo-v0.1",
+    similarity_fn_name=ScoringFunction.MAX_SIM,
+    use_instructions=True,
+    training_datasets=None,
```
Can you add training datasets to all models?
```suggestion
    training_datasets={"MMarcoReranking", "VDRMultilingualRetrieval"} | COLPALI_TRAINING_DATA,
```

dgolchin commented on mteb/models/model_implementations/slm_models.py:
```diff
@@ -0,0 +1,449 @@
+  archivePrefix={arXiv},
+  primaryClass={cs.IR}
+}
+"""
```
we used a forked version of colpali so we should mention the main project.

Samoed commented on mteb/models/model_implementations/slm_models.py:
```diff
@@ -0,0 +1,454 @@
+    similarity_fn_name=ScoringFunction.MAX_SIM,
+    use_instructions=True,
+    adapted_from="Qwen/Qwen3-VL-2B-Instruct",
+    training_datasets={"vidore/colpali_train_set"},
```
There should be mostly MTEB tasks. Please, change to
```suggestion
    training_datasets={"MMarcoReranking", "VDRMultilingualRetrieval"} | COLPALI_TRAINING_DATA,
```

Samoed commented on mteb/models/model_implementations/slm_models.py:
```diff
@@ -0,0 +1,454 @@
+}
+"""
+
+COLPALI_CITATION = """
```
Can you reuse citation from colpali models then? https://github.com/embeddings-benchmark/mteb/blob/480f1b99f54d193985dbb02ec2f8e8a55c5170a5/mteb/models/model_implementations/colpali_models.py#L192

Samoed commented on mteb/models/model_implementations/slm_models.py:
```diff
@@ -0,0 +1,454 @@
+    open_weights=True,
+    public_training_code=None,
+    public_training_data=None,
+    framework=["ColPali"],
```
You should add to all models
```suggestion
    framework=["ColPali"],
    model_type=["late-interactiion"],
```

KennethEnevoldsen commented on pyproject.toml:
```diff
@@ -93,6 +93,7 @@ nomic = ["einops>=0.8.1"]
 ark = ["volcengine-python-sdk[ark]==3.0.2", "tiktoken>=0.8.0"]
 colpali_engine = ["colpali_engine>=0.3.12"]
 colqwen3 = ["transformers>=4.57", "torchvision>=0.22.1"]
+sauerkrautlm-colpali = ["sauerkrautlm-colpali @ git+https://github.com/VAGOsolutions/sauerkrautlm-colpali.git"]
```
We may not want to rely directly on GitHub's main branch, as we don't have a guaranteed set of dependencies.
Would it be possible for you to do a versioned PyPI release?

dgolchin replied:
I added the git repo to pypi. The dependency is updated on the project.toml: https://pypi.org/project/sauerkrautlm-colpali/0.1.0/

KennethEnevoldsen reviewed (commented):
Hi thanks for the PR! Made a few minor suggestions in addition to @Samoed's. I think the github dependency might be a problem - would be great if we could get a pypi release.

We also generally report exact measures for number of parameters and memory usage. You can calculate memory usage using the `Metadata.calculate_memory_usage` once you have specified the number of parameters. This is to keep all of the metadata comparable.

dgolchin commented on mteb/models/model_implementations/slm_models.py:
```diff
@@ -0,0 +1,454 @@
+
+# =============================================================================
+# ColMinistral3 Model Metadata
+# =============================================================================
```
Thank you for your review! I removed the comments for consistency.

KennethEnevoldsen commented:
> Instead, the leaderboards usually rely on the size indicated in the model name (e.g., tomoro-colqwen3-embed-8b / -4b, ApsaraStackMaaS/EvoQwen2.5-VL-Retriever-7B-v1 / -3B-v1). We therefore used the same approach, but we can of course add the exact parameter sizes if that is required.

Made an issue on this https://github.com/embeddings-benchmark/mteb/issues/3811 - we should use the correct parameter count throughout.

Samoed commented on mteb/models/model_implementations/slm_models.py:
```diff
@@ -0,0 +1,470 @@
+
+
+=======
+>>>>>>> 32881a4 (fix: remove section headers and use PyPI package instead of Git URL)
```
Can you resolve conflicts?

dgolchin commented:
Thanks for the detailed review! I've addressed all the feedback:
I fixed the rebase conflict fragments that slipped through and removed the section header comments for consistency with the rest of the codebase. The sauerkrautlm-colpali library is now published on PyPI, so I updated the dependency to use the standard >=0.1.0 version constraint instead of the Git URL. I also calculated the exact n_parameters and memory_usage_mb values directly from the HuggingFace safetensors metadata for each model. The training_datasets field now properly imports COLPALI_TRAINING_DATA from colpali_models.py to maintain consistency.
Let me know if anything else needs to be adjusted!

KennethEnevoldsen commented:
thanks for finishing it up both of you - I think this is good to merge
